testsuites/basic/lsvmbus.py

```diff
@@ -47,42 +46,46 @@ class LsVmBus(TestSuite):
-            (num of vcpu/4, 64).
-        from the output of lsvmbus -vv command.
+        2. Check that each netvsc and storvsc SCSI device have correct number of vmbus
+            channels created and associated.
+            2.1 Check expected channel count of each netvsc is min (num of vcpu, 8).
+            2.2 Check expected channel count of each storvsc SCSI device is min (num of
+                 vcpu/4, 64).
         """,
         priority=1,
         requirement=simple_requirement(unsupported_os=[Windows]),
     )
-    def lsvmbus_channel_counting(self, environment: Environment, node: Node) -> None:
-        # get expected vm bus names
-        vmbus_class = VmbusNames("1" == node.tools[VmGeneration].get_generation())
-
+    def lsvmbus_count_devices_channels(
+        self, environment: Environment, node: Node
+    ) -> None:
+        # 1. Check expected vmbus device names presented in the lsvmbus output.
+        vmbus_devices = VmbusDeviceNames(
+            "1" == node.tools[VmGeneration].get_generation()
+        )
         lsvmbus_tool = node.tools[Lsvmbus]
-        vmbus_list = lsvmbus_tool.get_vmbuses()
-        actual_vmbus_names = [x.vmbus_name for x in vmbus_list]
-        assert_that(actual_vmbus_names).is_not_none()
-        assert_that(vmbus_class.names).is_subset_of(actual_vmbus_names)
+        vmbus_devices_list = lsvmbus_tool.get_device_channels_from_lsvmbus()
+        actual_vmbus_device_names = [x.name for x in vmbus_devices_list]
+        assert_that(actual_vmbus_device_names).is_not_none()
+        assert_that(vmbus_devices.names).is_subset_of(actual_vmbus_device_names)
 
-        # get actual core count
+        # 2. Check that each netvsc and storvsc SCSI device have correct number of
+        #  vmbus channels created and associated.
         lscpu_tool = node.tools[Lscpu]
         core_count = lscpu_tool.get_core_count()
         # Each netvsc device should have "the_number_of_vCPUs" channel(s)
-        # with a cap value of 8.
+        #  with a cap value of 8.
         expected_network_channel_count = min(core_count, 8)
-
         # Each storvsc SCSI device should have "the_number_of_vCPUs / 4" channel(s)
-        # with a cap value of 64.
+        #  with a cap value of 64.
         expected_scsi_channel_count = math.ceil(min(core_count, 256) / 4)
-
-        # When attach more than one nic
-        # channels of each nic need follow the same logic
-        for vmbus in vmbus_list:
-            if vmbus.vmbus_name == "Synthetic network adapter":
-                assert_that(vmbus.channel_vp_map).is_length(
-                    expected_network_channel_count
-                )
-            if vmbus.vmbus_name == "Synthetic SCSI Controller":
-                assert_that(vmbus.channel_vp_map).is_length(expected_scsi_channel_count)
+        for vmbus_device in vmbus_devices_list:
+            if vmbus_device.name == "Synthetic network adapter":
+                assert_that(
```

Comment:
The original assertion would be better. If you want to refine the output of `vmbus_device.channel_vp_map`, it needs to implement `__str__` of `ChannelVPMap`.

Reply:
vmbus.channel_vp_map is list of ChannelVPMap, I implemented __str__ of ChannelVPMap, but only vmbus.channel_vp_map[0].__str__() print expected value.